### PR TITLE
Change MessageListener

### PR DESCRIPTION
**Pull request by yemreinci**
Made the same changes I did on `EntryListener` and `ItemListener` to `MessageListener`. (see #604)

### CODE REVIEW

**Comment by yemreinci** [comment]
verify

**Comment by yemreinci** [comment]
verify

File: docsConfig [hunk 1]
```diff
@@ -784,7 +784,7 @@ EXCLUDE_SYMLINKS       = NO
 # Note that the wildcards are matched against the file with absolute path, so to
 # exclude all test directories for example use the pattern */test/*
 
-EXCLUDE_PATTERNS       = */util/* */iTest/* */test/* */collection/* */connection/* */executor/* */impl/* */map/* */multimap/* */protocol/* */proxy/* */queue/* */serialization/pimpl/* */spi/* */topic/* */txn/* */exception/pimpl/*
+EXCLUDE_PATTERNS       = */util/* */iTest/* */test/* */collection/* */connection/* */executor/* */impl/* */map/* */multimap/* */protocol/* */proxy/* */queue/* */serialization/pimpl/* */spi/* */txn/* */exception/pimpl/*
```
Comment (ihsandemir) [inline docsConfig:787]: i think this whole pattern needs to be updated. it looks out of date. we can do it in another PR. Can you open an issue to update this file?

File: examples/Org.Website.Samples/TopicSample.cpp [hunk 2]
```diff
@@ -17,20 +17,18 @@
...
 int main() {
     // Start the Hazelcast Client and connect to an already running Hazelcast Cluster on 127.0.0.1
     HazelcastClient hz;
     // Get a Topic called "my-distributed-topic"
     auto topic = hz.getTopic("my-distributed-topic");
     // Add a Listener to the Topic
-    topic->addMessageListener(TopicSample{}).get();
+    topic->addMessageListener(
+        topic::MessageListener().
+            on_received([](const topic::Message &message) {
```
Comment (ihsandemir) [inline examples/Org.Website.Samples/TopicSample.cpp:28]: previous API is passing as rvalue and I think we can keep it that way, since we will discard the object, better yet just pass it to the user so that he/she can move around without copy.

File: hazelcast/include/hazelcast/client/ITopic.h [hunk 3]
```diff
@@ -50,37 +53,26 @@ namespace hazelcast {
...
+            * Subscribe to this topic. 
+            * When a message is published on this topic, 
+            * the callback function that was provided to the given MessageListener object is called.
+            * \see MessageListener
             *
-            *      class MyMessageListener {
-            *      public:
-            *          //....
-            *
-            *          void onMessage(topic::Message<std::string> message) {
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/ITopic.h:62]: needs to be updated

File: hazelcast/include/hazelcast/client/topic/MessageListener.h [hunk 4]
```diff
@@ -28,27 +29,50 @@
...
-             * Invoked when a message is received for the added topic. Note that topic guarantees message ordering.
-             * Therefore there is only one thread invoking onMessage. The user should not keep the thread busy, but preferably
-             * should dispatch it via an Executor. This will increase the performance of the topic.
-             *
-             * @param message the message that is received for the added topic
+             * Listen to messages from an ITopic
+             * \see ITopic::addMessageListener
              */
-            class HAZELCAST_API MessageListener {
+            class HAZELCAST_API MessageListener final {
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/topic/MessageListener.h:41]: better naming suggestion: MessageListener -> listener

File: hazelcast/include/hazelcast/client/topic/MessageListener.h [hunk 5]
```diff
@@ -28,27 +29,50 @@
...
-                *
-                * @param message the message that is received for the added topic
-                */
-                virtual void onMessage(Message &&message) = 0;
-            };
+                 * \copydoc MessageListener::on_received
+                 */
+                template<typename Handler,
+                         typename = util::enable_if_rvalue_ref_t<Handler &&>>
+                MessageListener &&on_received(Handler &&h) && {
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/topic/MessageListener.h:64]: is this left-over? we already have this method above?

why are you returning rvalue here? I think just lvalue ref would be good enough. returning rvalue reference will also have its own problems https://stackoverflow.com/questions/1116641/is-returning-by-rvalue-reference-more-efficient

File: hazelcast/include/hazelcast/client/topic/MessageListener.h [hunk 6]
```diff
@@ -28,27 +29,50 @@
...
+                    on_received(std::move(h));
+                    return std::move(*this);
+                }
 
+            private:
+                using HandlerType = std::function<void(const Message &)>;
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/topic/MessageListener.h:70]: signature can take method parameter Message as rvalue.

File: hazelcast/include/hazelcast/client/topic/ReliableMessageListener.h [hunk 7]
```diff
@@ -60,8 +60,19 @@ namespace hazelcast {
...
+                virtual ~ReliableMessageListener() = default;
+
+                /**
+                * Invoked when a message is received for the added topic. Note that topic guarantees message ordering.
+                * Therefore there is only one thread invoking onMessage. The user should not keep the thread busy, but preferably
+                * should dispatch it via an Executor. This will increase the performance of the topic.
+                *
+                * @param message the message that is received for the added topic
+                */
+                virtual void onMessage(Message &&message) = 0;
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/topic/ReliableMessageListener.h:74]: i think this needs to match the signature of MessageListener on?received. Furthermore, lets try to remove the virtual methods where possible.

File: docsConfig (hunk 1 above)
Comment (yemreinci) [reply docsConfig:787]: Sure, I think we may also consider to review all documentation comments, doxygen gives lots of warnings. There are Javadoc things in the docs that are not supported or should be done differently in doxygen.

File: hazelcast/include/hazelcast/client/ITopic.h (hunk 3 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/ITopic.h:62]: will do

File: hazelcast/include/hazelcast/client/topic/MessageListener.h (hunk 5 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:64]: No, this is not sth accidental, I did the same thing on EntryListener and ItemListener as well. This is because we only accept rvalue parameters in addMessageListener, (or addItemListener and addEntryListener). If you don't return lvalue instead of rvalue ref to *this, then this won't compile:
```c++
topic.addMessageListener(
     MessageListener().
            on_receive(my_func)
)
```
But we and the compiler know that MessageListener() is rvalue here, and when on_receive is called, it will call the second overload with the `) && {`  and that will move the *this back to the caller as an rvalue. If we didnt have this overload than the user would have to use std::move explicitly to explicitly cast the returned lvalue back to rvalue.

Comment (yemreinci) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:64]: About the question on StackOverFlow, I don't think we are returning a dangling ref here, we are still returning *this not a copy of *this. The problem there would still exist if the question was asked with lvalues.

File: hazelcast/include/hazelcast/client/topic/ReliableMessageListener.h (hunk 7 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/topic/ReliableMessageListener.h:74]: I didn't wish to convert ReliableMessageListener on this PR, I just copy pasted the virtual methods from the old MessageListener and removed the inheritance so that it wouldn't break. I thought I would do it in a separate PR because ReliableMessageListener needs extra parameters other than handler functions (isLossTolerant, etc.) and I think we should discuss about how the API should look for those.

File: hazelcast/include/hazelcast/client/ITopic.h (hunk 3 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/ITopic.h:62]: I removed the example here and haven't written a new one, because from what I observed, this is not a common thing through the whole codebase (addEntryListener, addItemListener don't have examples as comments). And we already have an example under the `examples/` directory. 
As a suggestion for giving examples in the doc, doxygen has a `\example` command which we can use to give references to the examples in the `examples/` directory.

File: examples/Org.Website.Samples/TopicSample.cpp (hunk 2 above)
Comment (yemreinci) [reply examples/Org.Website.Samples/TopicSample.cpp:28]: But then it is not consistent with the other listeners, can we pass EntryEvent, ItemEvent etc. as rvalue?

Comment (ihsandemir) [reply examples/Org.Website.Samples/TopicSample.cpp:28]: We could, we are currently composing a separate event for each event message for each listener. If we are to make the event shared between listeners in the future, then it may be a different story but sharing events between listeners currently can only be considered for local listeners such as membership listener and such but for data structure listeners, the current way is a separate message for each listener even if they registered to the same data structure with the same flags.

File: hazelcast/include/hazelcast/client/ITopic.h (hunk 3 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/ITopic.h:62]: `example` tag is a good idea, will it include the examples in the doxygen generated pages?

Comment (yemreinci) [reply hazelcast/include/hazelcast/client/ITopic.h:62]: Yes, I guess it does. We need to set EXAMPLE_PATH.

File: hazelcast/include/hazelcast/client/topic/MessageListener.h (hunk 5 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:64]: can you give an example for the first method usage? From your answer i understand that only the second one would have been enough.

File: docsConfig (hunk 1 above)
Comment (yemreinci) [reply docsConfig]: Resetted this change, opened a new issue: https://github.com/hazelcast/hazelcast-cpp-client/issues/609.

Another question, is there a specific reason why the the config file is named docsConfig? Can we rename it to Doxyfile? It is the standard thing to do and one of the things that someone will look for when they first see the project.

File: hazelcast/include/hazelcast/client/topic/MessageListener.h (hunk 5 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:64]: This wouldn't compile, if we don't overload &:
```
MessageListener l;
l.on_receive(my_func);
topic_addMessageListener(std::move(l));
```
Well, this may look unlikely for the MessageListener, but when there are more events (EntryListener), someone might prefer this.

Comment (yemreinci) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:64]: This method of overloading both & and && might look like a code repeat, but it actually isn't. I am calling the & overloaded version from the &&. So, what you implemented in the & version, applies to the && version as well. && version will always be 2 lines: call the & version, and return rvalue ref to *this. Even if we needed to sth more complex in these setter methods (e.g. calculating listener flags in EntryListener), that thing would only be implemented once in the & overload, and the && would automatically use it.

File: hazelcast/include/hazelcast/client/topic/MessageListener.h (hunk 4 above)
Comment (yemreinci) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:41]: listener or Listener ?

File: examples/Org.Website.Samples/TopicSample.cpp (hunk 2 above)
Comment (yemreinci) [reply examples/Org.Website.Samples/TopicSample.cpp]: Ok, converted parameters back to `Message &&`. Created an issue for converting other events: https://github.com/hazelcast/hazelcast-cpp-client/issues/610

File: docsConfig (hunk 1 above)
Comment (ihsandemir) [reply docsConfig]: no reason but will need to update the release scripts when it is changed.

Comment (yemreinci) [reply docsConfig]: Ok, I am noting this in #609

File: hazelcast/include/hazelcast/client/topic/ReliableMessageListener.h (hunk 7 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/topic/ReliableMessageListener.h:74]: can you open an issue to correct this API?

File: hazelcast/include/hazelcast/client/topic/MessageListener.h (hunk 4 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/topic/MessageListener.h:41]: You can do `Listener` for now but I want to change all the code base to use all lower case, similar to the stl lib and boost lib naming.
